src/kernel.jl

```diff
@@ -84,14 +84,148 @@ function set_arg!(k::Kernel, idx::Integer, arg::LocalMem)
+        nextoffset = offset
+        for field in fieldnames(T)
+            xT = fieldtype(T, field)
+            nextoffset = _cl_sizeof(xT, nextoffset)
+        end
+        size = nextoffset - offset
+        size, size
+    end
+    offset = advance_aligned(offset, align)
+    offset += size
+    offset
+end
+
+Base.@generated function cl_sizeof{T}(x::T)
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+Base.@generated function cl_sizeof{T}(x::Type{T})
+    size = _cl_sizeof(T, 0)
+    :($size)
+end
+
+@generated function aligned_convert(x)
+    offset = Ref(0); elements = []; fields = []
+    _aligned_convert!(x, offset, elements, fields, :x)
+    ret = if length(elements) == 1 # no conversion happened
+        return :(x, $(offset[]))
+    else
+        tupl = Expr(:tuple)
+        tupl.args = first.(elements)
+        expr = quote
+            $(fields...) # hoisted field loads
+            $tupl, $(offset[])
+        end
+        expr
+    end
+end
+
+@generated function aligned_convert{T}(x::Type{T})
+    offset = Ref(0); elements = []; fields = []
+    _aligned_convert!(T, offset, elements, fields, :x)
+    ret = if length(elements) == 1 # no conversion happened
+        return :(T, $(offset[]))
+    else
+        expr = quote
+            Tuple{$(last.(elements)...)}, $(offset[])
+        end
+        expr
+    end
+end
+
+function _aligned_convert!(x, offset = Ref(0), elements = [], fields = [], fieldname = gensym(:field))
+    alignment = cl_alignement(x)
+    if alignment != 0 && offset[] % alignment != 0
+        npad = ((div(offset[], alignment) + 1) * alignment) - offset[]
+        pad = Pad{npad}()
+        offset[] += npad
+        push!(elements, :(Pad{$npad}()) => Pad{npad})
+    end
+    if !is_cl_inbuild(x) && nfields(x) > 0
+        for field in fieldnames(x)
+            current_field = gensym(string(field))
+            push!(fields, :($current_field = getfield($fieldname, $(QuoteNode(field)))))
+            xelem = fieldtype(x, field)
+            _aligned_convert!(xelem, offset, elements, fields, current_field)
+        end
+    else
+        push!(elements, fieldname => x)
+        offset[] += sizeof(x)
+    end
+    return
+end
+
+
+
 #TODO: type safe calling of set args for kernel (with clang)
+# is 1024 a good number?
+const _arg_tmp_buffer = Vector{Int8}(1024)
 function set_arg!{T}(k::Kernel, idx::Integer, arg::T)
     @assert idx > 0 "Kernel idx must be bigger 0"
     if !isbits(T) # TODO add more thorough mem layout checks and the clang stuff
         error("Only isbits types allowed. Found: $T")
     end
-    boxed_arg = Ref{T}(arg)
-    @check api.clSetKernelArg(k.id, cl_uint(idx - 1), sizeof(T), boxed_arg)
+    aligned_arg, off = aligned_convert(arg)
+    if sizeof(aligned_arg) > 1024
+        error("Can't upload single types to opencl that are bigger than 1024 bytes")
+    end
+    T_aligned = typeof(aligned_arg)
+    # is this save?
+    unsafe_store!(Ptr{T_aligned}(pointer(_arg_tmp_buffer)), aligned_arg)
```

Comment:
Because the syntax was added long before stack allocation of `Ref` is brought up and before `RefValue`/`Ref` is even a type. It was never added as an optimization so it was never a premature optimization by definition. Also, without knowing Base is actually ready for stack allocation of `Ref` as well as non-experimental multithreading for more and more cases this won't be a premature optimization either. The correct way to do thing changes so now it is definitely a premature optimization.